xwiki-platform-core/xwiki-platform-rendering/xwiki-platform-rendering-macros/xwiki-platform-rendering-macro-code/src/main/java/org/xwiki/rendering/internal/macro/code/source/CodeMacroSourceReferenceConverter.java

```diff
@@ -0,0 +1,134 @@
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import org.apache.commons.configuration2.ex.ConversionException;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang3.reflect.TypeUtils;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.component.manager.ComponentLookupException;
+import org.xwiki.component.manager.ComponentManager;
+import org.xwiki.properties.converter.AbstractConverter;
+import org.xwiki.rendering.macro.code.source.CodeMacroSourceReference;
+
+/**
+ * Convert various types (especially String) to {@link CodeMacroSourceReference}.
+ * 
+ * @version $Id$
+ * @since 15.0RC1
+ * @since 14.10.2
+ */
+@Component
+@Singleton
+public class CodeMacroSourceReferenceConverter extends AbstractConverter<CodeMacroSourceReference>
+{
+    @Inject
+    @Named("context")
+    private Provider<ComponentManager> contextComponentManagerProvider;
+
+    @Override
+    protected CodeMacroSourceReference convertToType(Type targetType, Object value)
+    {
+        if (value == null) {
+            return null;
+        }
+
+        try {
+            return convertToType(value);
+        } catch (IOException e) {
+            throw new ConversionException("Failed to convert to CodeMacroSourceReference", e);
+        }
+    }
+
+    private CodeMacroSourceReference convertToType(Object value) throws IOException
+    {
+        CodeMacroSourceReference reference;
+
+        if (value instanceof String) {
+            reference = fromString((String) value);
+        } else if (value instanceof InputStream) {
+            reference = new CodeMacroSourceReference(CodeMacroSourceReference.TYPE_STRING,
+                IOUtils.toString((InputStream) value, StandardCharsets.UTF_8));
+        } else if (value instanceof byte[]) {
+            reference = new CodeMacroSourceReference(CodeMacroSourceReference.TYPE_STRING,
+                new String((byte[]) value, StandardCharsets.UTF_8));
+        } else if (value instanceof Reader) {
+            reference =
+                new CodeMacroSourceReference(CodeMacroSourceReference.TYPE_STRING, IOUtils.toString((Reader) value));
+        } else if (value instanceof URL) {
+            reference = new CodeMacroSourceReference(CodeMacroSourceReference.TYPE_URL, ((URL) value).toExternalForm());
+        } else if (value instanceof File) {
+            reference =
+                new CodeMacroSourceReference(CodeMacroSourceReference.TYPE_FILE, ((File) value).getAbsolutePath());
+        } else {
+            reference = fromUnknownType(value);
+        }
+
+        return reference;
+    }
+
+    private CodeMacroSourceReference fromUnknownType(Object value)
+    {
+        ParameterizedType componentRole = TypeUtils.parameterize(
+            org.xwiki.rendering.macro.code.source.CodeMacroSourceReferenceConverter.class, value.getClass());
```

Comment:
Why is this type not imported?

Reply:
Because the class name is the same than the current class.